### PR TITLE
Make textfields restorable

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

Enabled the TextField widgets (Material's `TextField`, `CupertinoTextField`) to restore their scroll offset and their text content with the state restoration framework.

This PR requires https://github.com/flutter/flutter/pull/63131 to land first.

## Related Issues

#62916

## Tests

I added the following tests:

* Restoration tests for `TextField` and `CupertinoTextField`.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by goderbauer** [comment]
Tests will fail until https://github.com/flutter/flutter/pull/63131 is merged and this PR is rebased.

File: packages/flutter/test/cupertino/text_field_restoration_test.dart [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            ),
+          ),
+        ),
+      ),
+    );
+  }
+}
+
+Future<void> skipPastScrollingAnimation(WidgetTester tester) async {
+  await tester.pump();
```
Comment (justinmc) [inline packages/flutter/test/cupertino/text_field_restoration_test.dart:124]: Is this pump necessary even with the duration pump on the next line?

Same question for the material text field test.

File: packages/flutter/test/material/textfield_restoration_test.dart [hunk 2]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Comment (justinmc) [inline packages/flutter/test/material/textfield_restoration_test.dart:1]: The name of this file is missing an underscore, should be "text_field_restoration_test".

**Review by justinmc — APPROVED** [review]
LGTM besides a small naming typo in my comment below.  I'm really excited for the ability to restore text fields, thanks for doing this!

I also wanted to make sure that you've thought about TextFormField as well.  It just renders a TextField, so it should pick up all of the restoration logic here, but what about say its validation state? If a TextFormField is invalid and displaying an error, and then later it is restored, will it still be showing the validation error message?  Maybe that's a separate issue.

File: packages/flutter/test/material/textfield_restoration_test.dart (hunk 2 above)
Comment (goderbauer) [reply packages/flutter/test/material/textfield_restoration_test.dart:1]: Good catch. Fixed.

File: packages/flutter/test/cupertino/text_field_restoration_test.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/test/cupertino/text_field_restoration_test.dart:124]: We need one pump to start the animation. And another (with a duration) to finish it.

**Comment by goderbauer** [comment]
> I also wanted to make sure that you've thought about TextFormField as well.

I've not looked at that one yet, but added it to the backlog in https://github.com/flutter/flutter/issues/62886.

**Pull request merged**